crates/ruff_db/src/file_system.rs

```diff
@@ -0,0 +1,265 @@
+use std::fmt::Formatter;
+use std::ops::Deref;
+use std::path::Path;
+
+use camino::{Utf8Path, Utf8PathBuf};
+use filetime::FileTime;
+
+pub use memory::MemoryFileSystem;
+pub use os::OsFileSystem;
+
+mod memory;
+mod os;
+
+pub type Result<T> = std::io::Result<T>;
+
+/// A file system that can be used to read and write files.
+///
+/// The file system is agnostic to the actual storage medium, it could be a real file system, a combination
+/// of a real file system and an in-memory file system in the case of an LSP where unsaved changes are stored in memory,
+/// or an all in-memory file system for testing.
+pub trait FileSystem {
+    /// Reads the metadata of the file or directory at `path`.
+    fn metadata(&self, path: &FileSystemPath) -> Result<Metadata>;
+
+    /// Reads the content of the file at `path`.
+    fn read(&self, path: &FileSystemPath) -> Result<String>;
+
+    /// Returns `true` if `path` exists.
+    fn exists(&self, path: &FileSystemPath) -> bool;
+}
+
+// TODO support untitled files for the LSP use case. Wrap a `str` and `String`
+//    The main question is how `as_std_path` would work for untitled files, that can only exist in the LSP case
+//    but there's no compile time guarantee that a [`OsFileSystem`] never gets an untitled file path.
+
+/// Path to a file or directory stored in [`FileSystem`].
+///
+/// The path is guaranteed to be valid UTF-8.
+#[repr(transparent)]
+#[derive(Eq, PartialEq, Hash)]
+pub struct FileSystemPath(Utf8Path);
+
+impl FileSystemPath {
+    pub fn new(path: &(impl AsRef<Utf8Path> + ?Sized)) -> &Self {
+        let path = path.as_ref();
+        unsafe { &*(path as *const Utf8Path as *const FileSystemPath) }
+    }
+
+    /// Converts the path to an owned [`FileSystemPathBuf`].
+    pub fn to_path_buf(&self) -> FileSystemPathBuf {
+        FileSystemPathBuf(self.0.to_path_buf())
+    }
+
+    /// Returns the path as a string slice.
+    pub fn as_str(&self) -> &str {
+        self.0.as_str()
+    }
+
+    /// Returns the std path for the file.
+    pub fn as_std_path(&self) -> &Path {
+        self.0.as_std_path()
+    }
+}
+
+/// Owned path to a file or directory stored in [`FileSystem`].
+///
+/// The path is guaranteed to be valid UTF-8.
+#[repr(transparent)]
+#[derive(Eq, PartialEq, Clone, Hash)]
+pub struct FileSystemPathBuf(Utf8PathBuf);
+
+impl Default for FileSystemPathBuf {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl FileSystemPathBuf {
+    pub fn new() -> Self {
+        Self(Utf8PathBuf::new())
+    }
+
+    pub fn as_path(&self) -> &FileSystemPath {
+        // SAFETY: FsPath is marked as #[repr(transparent)] so the conversion from a
+        // *const Utf8Path to a *const FsPath is valid.
+        unsafe { &*(self.0.as_path() as *const Utf8Path as *const FileSystemPath) }
+    }
+}
+
+impl AsRef<FileSystemPath> for FileSystemPathBuf {
+    fn as_ref(&self) -> &FileSystemPath {
```

Comment:
Probably most methods in here should be marked as `inline`. Not that I think it matters too much for our use case. Inline tells the Rust compiler to emit sufficient information that cross-crate inlining works even if LTO is disabled. That's less important for us because we have LTO (thin) enabled. 

My general thinking is: If the method is only a very thin wrapper, that is called frequently add `[#inline]`. But I'm not super consistent on it as you can see